h2o-algos/src/main/java/hex/svm/PrimalDualIPM.java

```diff
@@ -0,0 +1,502 @@
+    } finally {
+      volatileWorkspace.remove();
+    }
+  }
+
+  private static Vec solve(Frame rbicf, Vec label, Params params, Frame volatileWorkspace) {
+    Frame workspace = new Frame(new String[]{"label"}, new Vec[]{label});
+    workspace.add("x", label.makeZero());
+    workspace.add(volatileWorkspace);
+    
+    new InitTask(params).doAll(workspace);
+
+    Vec z = workspace.vec("z");
+    Vec la = workspace.vec("la");
+    Vec xi = workspace.vec("xi");
+    Vec x = workspace.vec("x");
+    Vec dxi = workspace.vec("dxi");
+    Vec dla = workspace.vec("dla");
+    Vec d = workspace.vec("d");
+    Vec dx = workspace.vec("dx");
+
+    double nu = 0;
+    boolean converged = false;
+    final long num_constraints = rbicf.numRows() * 2;
+    for (int iter = 0; iter < params._max_iter; iter++) {
+      final double eta = new SurrogateGapTask(params).doAll(workspace)._sum;
+      final double t = (params._mu_factor * num_constraints) / eta;
+      Log.debug("sgap: " + eta + " t: " + t);
+
+      computePartialZ(rbicf, x, params._tradeoff, z);
+      CheckConvergenceTask cct = new CheckConvergenceTask(params, nu).doAll(workspace);
+      converged = cct._resp <= params._feasible_threshold && cct._resd <= params._feasible_threshold && eta <= params._sgap_bound;
+      if (converged) {
+        break;
+      }
+
+      new UpdateVarsTask(params, t).doAll(workspace);
+
+      LLMatrix icfA = MatrixUtils.productMtDM(rbicf, d);
+      icfA.addUnitMat();
+      LLMatrix lra = icfA.cf();
+
+      final double dnu = computeDeltaNu(rbicf, d, label, z, x, lra);
+      computeDeltaX(rbicf, d, label, dnu, lra, z, dx);
+      
+      LineSearchTask lst = new LineSearchTask(params).doAll(workspace);
+
+      new MakeStepTask(lst._ap, lst._ad).doAll(x, dx, xi, dxi, la, dla);
+
+      nu += lst._ad * dnu;
+    }
+
+    if (! converged) {
+      Log.warn("The algorithm didn't converge in the maximum number of iterations. " +
+              "Please consider changing the convergence parameters or increase the maximum number of iterations (" + params._max_iter + ").");
+    }
+
+    volatileWorkspace.remove();
+    
+    return x;
+  }
+
+  private static abstract class PDIPMTask<E extends PDIPMTask<E>> extends MRTask<E> {
+    transient Chunk _label, _x;
+    transient Chunk _z;
+    transient Chunk _xi, _dxi, _la, _dla;
+    transient Chunk _tlx, _tux, _xilx, _laux, _d;
+    transient Chunk _dx;
+
+    final double _c_pos;
+    final double _c_neg;
+
+    PDIPMTask(Params params) {
+      _c_pos = params._weight_positive * params._hyper_parm;
+      _c_neg = params._weight_negative * params._hyper_parm;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      _label = cs[0];
+      _x = cs[1];
+      _z = cs[2];
+      _xi = cs[3];
+      _dxi = cs[4];
+      _la = cs[5];
+      _dla = cs[6];
+      _tlx = cs[7];
+      _tux = cs[8];
+      _xilx = cs[9];
+      _laux = cs[10];
+      _d = cs[11];
+      _dx = cs[12];
+      map();
+    }
```

Comment:
Michalk:

I know what most of your variable names mean except the following:

_tlx, _tux, _xilx, _laux.  Can you tell me what they are?

Reply:
These temporary vectors and they are not referenced in the paper, I took them from the reference implementation:

    //     tlx = \frac{1}{t \alpha}
    //     tux = \frac{1}{t (C - \alpha)}
    //     xilx = \frac{\xi}{\alpha}
    //     laux = \frac{\lambda}{C - \alpha}
    //     D^(-1) = \diag(\frac{\xi}{\alpha} + \frac{\lambda}{C - \alpha})

Reply:
I finally got it after going through your UpdateVarsTask!